### DISCUSSION TITLE
How to reverse engineer an bluetooth HCI log to add a new Printer ?

### DISCUSSION THREAD

**Discussion by anthony-foulfoin** [bitrate16/ppa6-python#18] [Q&A] [answered]
Hi,

I have a Peripage A40+ printer 304 dpi. I wanted to make a PR to add its support, but even with all the bluetooth logs and analysing in wireshark I don't know where to find the necessary informations. 
I looked at the commit that added the A40 [https://github.com/bitrate16/peripage-python/pull/11/commits/90683dcd3ffb53c42639473c81036fe4463e82ed](url) but it doesn't help me to know what to search for in the logs. Where do you find all those informations ?

Thanks !

**Comment by bitrate16** [comment, marked as answer]
Try this code:
```python
import peripage

printer = peripage.Printer(
    # MAC
    mac='00:15:83:15:BC:5F',
    # Doesn't matter:
    printer_type=peripage.PrinterType.A6p
)
printer.connect()

# Print many letters in hex to determine the width
printer.writeASCII('0123456789ABCDEF' * 20)

printer.disconnect()
```

This shoud print `0123456789ABCDEF` repeated 20 times. Count length of line in 16base, convert to 10base and you will get the width of line in characters.

Next, add your printer spec here:
https://github.com/bitrate16/peripage-python/blob/main/peripage/__init__.py#L46
similar to others, or tell me and i'll add:
```python
row_characters=SOME_VALUE_YOU_GOT
row_width=row_characters * 12
row_bytes=row_width // 8
```
